Add information about RFC process (#1)
valkey-io/valkey-rfc: README.md

Valkey RFC
==========

> This is a collection of feature proposals and descriptions of changes to Valkey

Reviewer: ```suggestion
This repository is a collection of feature proposals and descriptions of changes to Valkey
```

> that require some more details than just the text in a pull request or an issue.

Reviewer: ```suggestion
that require some more detail than just the text in a pull request or an issue.
```

It is loosely inspired by RFCs and by Python's enhancement proposals (PEP).

Each feature or larger topic is described in a markdown file that's named in
uppercase and ends in `.md`. These files are not formally numbered, but we use

> the pull request number that added an RFC to refer to the change. For example,

Reviewer: ```suggestion
the pull request number that initially added an RFC to refer to the change. For example,
```

this description in the README.md file was written in RFC #1.

Workflow
--------

An RFC starts off as a pull request. It's reviewed for formatting, style and

> consistency. Then the proposal is merged. This doesn't mean that the feature is

Reviewer: I think I would also include there is a standard for quality of an RFC on the content. If something is very vague and unclear, it's probably worth improving it before merging it. We won't reject it based off the proposal though.

approved for inclusion in Valkey. It's still just a proposal.

Each file has a status like "draft", "suggested", "approved" or "rejected". The

> possible statuses are yet to be decided.

Reviewer: I think we should decide that now. "Draft", "Published", "Approved", and "Rejected".

Author: Yes. So let's define what these mean and choose terms that are unambigous.

- Draft = non yet merged PR (like this one). I don't think there's any reason to actually spell this out in the document itself. It's enough that the PR is not yet merged.
- Published = RFC merged to this repo but no decision made. PEPs seem to use Draft for this and they accept changes in this stage. I think Published can be interpreted as if the feature has been published, but it's just the RFC. Can we find a better term for this? I propose "Proposed". This is the stage where most of the discussion happens, as in Request For Comments. Perhaps we don't need to spell out a status at all. An RFC without a status is implicitly in this status.
- Approved = Specification approved by core team, awaiting implementation. Minor changes can still be requested in the actual implementation PR. (PEPs use "Accepted" for this.)
- Rejected = Not going to be implemented.

The PEP process has one more status "Final", meaning the feature has been implemented, merged, released. We should update an RFC to mention in which Valkey version it was released and add links to the implementation PR.

Reviewer: Should we add the list of states and what they mean here? Right now it's unclear what informational is unless you go look it up in the official IETF documentation.

Author: I ended up just adding Proposed, Approved and Rejected. I imagined they can't be misunderstood, but they probably can. I'll define them.

The core team can later change the status and make changes. For larger changes,
the PR making the change is mentioned too and can be referred to by their
respective pull-request numbers.

What's useful to include?
-------------------------

This is not a strict requirement of what to include in each file. What's useful

> to include may vary from case to case.

Reviewer: Why not make it strict?

Author: We want it to have less friction, be less formal than PEPs and RFCs right?

I think we can require this content but we don't have to enforce a specific order and headings within the document IMO. If somebody already has a document (like RDMA.md, TRIGGERS.md, CLUSTER.md) it can be added without too much rewrite.

I'll rewrite this part.

Reviewer: I guess I would just like to have a strict set of stuff you "must" have, but give a lot of flexibility in those sections.

Author: Please suggest a formulation.

* Status and the numbers of the PRs adding or making major changes to the file.
* Abstract. A few sentences describing the feature.
* Motivation. What the feature solves and why the existing functionality is not
  enough.

> * Rationale. Why certain design decisions have been made. Comparisons with
>   similar features in other projects.
> * Specification. A more detailed description of the feature.

Reviewer: ```suggestion
* Design considerations. A description of the design constraints and requirements
  for the proposal.
* Specification. A more detailed description of the feature including 
  why certain design decisions have been made. Comparisons with
  similar features in other projects.
```
A slightly different wording to emphasize the constraints and incorporate the decision making inside the specification.

Author: Why not keep the specification concise and separate out the "why"?

Small comparisons may make sense to incorporate in the specification, but larger ones may be better to move to its own section or subsection.

Reviewer: I'm ok with what you ended up.

> * Links to related issues and PRs on this and other repos.

Reviewer: ```suggestion
* Links to related material such as issues, pull requests, papers, or other references.
```

that require some more detail than just the text in a pull request or an issue.
It is loosely inspired by RFCs and by Python's enhancement proposals (PEP).

Each feature or larger topic is described in a markdown file that's named in
uppercase and ends in `.md`. These files are not formally numbered, but we use
the pull request number that initially added an RFC to refer to the change. For example,
this description in the README.md file was written in RFC #1.

Workflow
--------

An RFC starts off as a pull request. It's reviewed for formatting, style,
consisteny and content quality. The content shouldn't be very vague or unclear.
Then the proposal is merged. This doesn't mean that the feature is approved for
inclusion in Valkey. It's still just a proposal.

Each file has a status "Proposed", "Approved" or "Rejected" (or "Informational"
like this README file).

The core team can later change the status and make changes. For larger changes,
the PR making the change is mentioned too and can be referred to by their
respective pull-request numbers.

What's useful to include?
-------------------------

> Please include the following content. What information is useful to include may
> vary from case to case, but this is a guideline.

Reviewer: ```suggestion
The Valkey RFC format is not a strict format, but should include the following section unless they are unnecessary for the proposal you are submitting.
```

Just to say, you should use this unless you have a reason not to use this.

> ---

Reviewer: Can you comment about this.
